### PR TITLE
Issue #3211: Fix NPE when creating schema after deleting a schema

### LINKED ISSUE #2786: Problems introducing incompatible changes to persistent topic schema
#### Expected behavior

When introducing incompatible changes to the schema for a persistent topic, deleting the schema/topic should allow the new schema to be used.

#### Actual behavior

After deleting the schema using pulsar-admin, the Java client gets an NPE when trying to create the producer. The same NPE happens if you delete schema, topic, namespace and tenant, then recreate tenant and namespace.

#### Steps to reproduce
I guess steps 2 and 3 aren't actually necessary to reproduce this, but I've included everything I did just in case:
1. Run a producer using a JSON schema - we're using Kotlin, but I'm pretty confident the same would apply in Java. These are the relevant snippets:
```kotlin
data class Test(var fieldOne: String = UNKNOWN,
                var fieldTwo: String = UNKNOWN) : Serializable

val producer = client.newProducer(JSONSchema.of(Test::class.java))
          .topic("persistent://hicksst_hicksco_qa/default/TestTopic")
          .createAsync()
          .get(30L, TimeUnit.SECONDS)
```
2. Make an incompatible change to the schema:
```kotlin
data class Test(var fieldOne: String = UNKNOWN,
                var fieldTwo: Int = 0) : Serializable
```
3. Running the producer and consumer now both throw an "incompatible schema" exception as expected.

4. Delete the schema using pulsar-admin:

```
pulsar-admin schemas delete persistent://hicksst_hicksco_qa/default/TestTopic
```
5. Running the producer now throws the following on the call to CompletableFuture.get:
```
Caused by: java.util.concurrent.ExecutionException: org.apache.pulsar.client.api.PulsarClientException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at percentile.streams.trade.simulator.pulsar.PulsarProducer.initialise(PulsarProducer.kt:34)
	... 5 more
Caused by: org.apache.pulsar.client.api.PulsarClientException: java.lang.NullPointerException
	at org.apache.pulsar.client.impl.ClientCnx.getPulsarClientException(ClientCnx.java:761)
	at org.apache.pulsar.client.impl.ClientCnx.handleError(ClientCnx.java:498)
	at org.apache.pulsar.common.api.PulsarDecoder.channelRead(PulsarDecoder.java:154)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at org.apache.pulsar.shade.io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at org.apache.pulsar.shade.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at org.apache.pulsar.shade.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1414)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at org.apache.pulsar.shade.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:945)
	at org.apache.pulsar.shade.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:146)
	at org.apache.pulsar.shade.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
	at org.apache.pulsar.shade.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
	at org.apache.pulsar.shade.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
	at org.apache.pulsar.shade.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
	at org.apache.pulsar.shade.io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
	at org.apache.pulsar.shade.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```
There are no exceptions shown in the Pulsar server logs. 
Running the consumer does not throw an exception.

6. Deleting and recreating the tenant, then running the producer throws the same NPE.

#### System configuration
**Pulsar version**: 2.1.0
**Java client library version**: 2.1.1-incubating

### LINKED ISSUE #3211: pulsar producer create fail, get error :Received error from server: java.lang.NullPointerException
#### Actual behavior

I have create a producer with JSON schema of Person object for topic named /xh/test/demo8, in the code review, we need to update Person to Student, the Student extends Person, so i delete the schema for persistent://xh/test/demo8, then i create the producer by pulsarClient, but the server response the error Received error from server: java.lang.NullPointerException. Then, i delete the topic persistent://xh/test/demo8, and create producer ,i get a same error from server, so that, I use pulsar-admin to upload a new schema by conf file, the server response a 500 bad code for me

#### Steps to reproduce

1、create a producer with Json schema with a pojo object
2、delete the topic schema
3、create producer 

#### System configuration
**Pulsar version**: 2.2.0

### PR DESCRIPTION
**Pull request by sijie**
Fixes #3211
Fixes #2786 

*Motivation*

When a schema is deleted, the schema is not removed directly.
You can still fetch the latest schema but its state is marked as `deleted`.

So when we apply schema compatability check, we should ignore deleted schema.

*Modifications*

Ignore deleted scheam when doing schema compatability check

### CODE REVIEW

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/Schemas.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+public final class Schemas {
+
+    /**
+     * A Person Struct.
+     */
+    @Data
+    @Getter
+    @Setter
+    @ToString
+    @EqualsAndHashCode
```
Comment (merlimat) [inline tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/Schemas.java]: nit: `@Data` is already a shortcut for all the other annotations

**Comment by merlimat** [comment]
run integration tests

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/Schemas.java (hunk 1 above)
Comment (sijie) [reply tests/integration/src/test/java/org/apache/pulsar/tests/integration/schema/Schemas.java]: lol. I copied this pattern from some other avro schema test files :)

**Comment by sijie** [comment]
run integration tests

**Comment by sijie** [comment]
run integration tests

**Comment by codelipenghui** [comment]
run integration tests

**Comment by merlimat** [comment]
Merged in 2.3.1 at 47f85a1

**Pull request merged**